File: python/protein-translation/protein_translation.py
protein_list = {
    "AUG": "Methionine",
    "UUU": "Phenylalanine",
    "UUC": "Phenylalanine",
    "UUA": "Leucine",
    "UUG": "Leucine",
    "UCU": "Serine",
    "UCC": "Serine",
    "UCA": "Serine",
    "UCG": "Serine",
    "UAU": "Tyrosine",
    "UAC": "Tyrosine",
    "UGU": "Cysteine",
    "UGC": "Cysteine",
    "UGG": "Tryptophan"
}

def proteins(strand):
    codons = []
    result = []
    for i in range(0, len(strand), 3):
        codons.append(strand[i: i + 3])
    for i in range(len(codons)):
        protein = protein_list.get(codons[i])
        if protein:
            result.append(protein)
        else:
            break
    return result

File: javascript/protein-translation/protein-translation.js
const proteins = {
    Methionine: ["AUG"],
    Phenylalanine: ["UUU", "UUC"],
    Leucine: ["UUA", "UUG"],
    Serine: ["UCU", "UCC", "UCA", "UCG"],
    Tyrosine: ["UAU", "UAC"],
    Cysteine: ["UGU", "UGC"],
    Tryptophan: ["UGG"],
    STOP: ["UAA", "UAG", "UGA"],
};

const getProtein = (codon) => {
    return Object.keys(proteins).find((x) => proteins[x].includes(codon));
};

export const translate = (strand = "") => {
    let res = [];
    for (var i = 0; i < strand.length; i += 3) {
        let protein = getProtein(strand.substring(i, i + 3));
        if (!protein) throw new Error("Invalid codon");
        if (protein === "STOP") break;
        res.push(protein);
    }
    return res;
};
